pkg/qrcode/lora_alliance_tr005d3.go

```diff
@@ -0,0 +1,175 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package qrcode
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"strings"
+
+	pbtypes "github.com/gogo/protobuf/types"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+)
+
+// LoRaAllianceTR005Draft3 is the LoRa Alliance defined format in Technical Recommendation TR005 Draft 3.
+type LoRaAllianceTR005Draft3 struct {
+	JoinEUI,
+	DevEUI types.EUI64
+	VendorID [2]byte
+	ModelID  [2]byte
+	DeviceValidationCode,
+	SerialNumber,
+	Proprietary string
+}
+
+// Encode implements the Data interface.
+func (m *LoRaAllianceTR005Draft3) Encode(dev *ttnpb.EndDevice) error {
+	if dev.JoinEUI == nil {
+		return errNoJoinEUI
+	}
+	if dev.DevEUI == nil {
+		return errNoDevEUI
+	}
+	*m = LoRaAllianceTR005Draft3{
+		JoinEUI:              *dev.JoinEUI,
+		DevEUI:               *dev.DevEUI,
+		DeviceValidationCode: dev.GetClaimAuthenticationCode().GetValue(),
+	}
+	return nil
+}
+
+// validTR005Draft3ExtensionChars defines only alphanumeric characters.
+const validTR005Draft3ExtensionChars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+func (m LoRaAllianceTR005Draft3) validateExtensionChars(s string) error {
+	for _, r := range s {
+		if strings.IndexRune(validTR005Draft3ExtensionChars, r) == -1 {
+			return errCharacter.WithAttributes("r", r)
+		}
+	}
+	return nil
+}
+
+// Validate implements the Data interface.
+func (m LoRaAllianceTR005Draft3) Validate() error {
```

Comment:
FTR, this is not used by generated validators. It may make sense to implement `ValidateFields(...string) error` as well. `Validate() error` is then just `func() error { return ValidateFields(); }`

Reply:
There are no fields here